openedx/core/djangoapps/content_libraries/docs/decisions/0004-upstream-block.rst

```diff
@@ -0,0 +1,89 @@
+4. Persist the relationship between blocks and their upstream source
+####################################################################
+
+Status
+******
+
+Accepted
+
+Pending implementation in (TODO - link)
+
+Context
+*******
+
+* TODO - context
+
+Decision
+********
+
+* The following two fields will be added to every XBlock in CMS:
+
+  .. code-block:: python
+
+     upstream_block = String(
+         scope=Scope.settings,
+         help=(
+             "The usage key of a the block that served as this block's template, if one exists. "
+             "Generally, the upstream_block is within a Content Library."
+         ),
+         hidden=True,
+         default=None,
+         ...
+     )
+     upstream_block_version = Integer(
+         scope=Scope.settings,
+         help=(
+             "The upstream_block's version number, at the time this block was created from it. "
+             "If this version is older than the upstream_block's latest version, then CMS will "
+             "allow this block to fetch updated content from upstream_block."
+         ),
+         hidden=True,
+         default=None,
+         ...
+     )
+
+* The upstream-defined values of XBlock settings will be exposed in the XBlock API (TODO - details)
+
+* The upstream-defined values will be set when the XBlock is initialized by (TODO - figure this out)
+
+* The existing ``get_block_original_usage`` Modulestore methods will (TODO - figure this out)
+
+* New attributes will be serialized into the OLX of blocks with upstreams:
+
+  * The two new XBlock fields will serialize into OLX as attributes, like any settings-scoped field.
+  * The upstream defaults will use a new special syntax: ``upstream.$FIELD_NAME``.
+  * The period (``.``) was explicitly chosen to ensure that these special attributes cannot possibly conflict with any Python-defined XBlock fields.
+
+    .. code-block:: xml
+
+       <problem
+         display_name="A title that has been customized in the course"
+         max_attempts=2
+         upstream_block="lb:myorg:mylib:problem:p1"
+         upstream_block_version="12"
+         upstream_block.display_name="The title that was defined in the library block"
+         upstream_block.max_attempts=3
+       >
+       <!-- problem content would go here -->
+       </problem>
+
+* Existing LibraryContentBlock children will be missing these attributes, and we will need to handle that (TODO - more details)
+
+Consequences
+************
```

Comment:
I think there's an extra set of consequences here for defaults and inheritance:

* We'll need to modify the XBlock runtime so that it grabs value in the order: explicitly set on block, inherited via parent, upstream.
* We should be able to (eventually) remove the [special case handling](https://github.com/openedx/edx-platform/blob/09a8cfcccfca0a26b7b75e723587db386658b2c7/xmodule/modulestore/inheritance.py#L386-L393) that InheritingFieldData makes for `library_content`. Honestly, there's so much weirdness around how LibraryContent currently work that I'd rather make a new block type entirely rather than bridging the old stuff and have something external to the two do the conversion.

Reply:
I'm also going to go out on a limb and say that `upstream.{}` tags are _not_ inheritable. 😛

Reply:
Another thing I thought of: If we do this, I think it would be really valuable to ensure that [v2] library exports encode the full library ID in the export .tar.gz file. Currently, the convention is that Open edX exports don't include the ID of the thing you're exporting, and in fact you usually have to create an empty course/library with a new ID before you import. But in the case of libraries, we're going to want to encourage situations where IDs match across instances, and I'm thinking that a good way to do that is to include the full ID of the library and change the import workflow so that by default when you import a library to another instance, it has the same ID, and thus the `upstream_block` references will be valid. Of course, it won't break much if they're not valid, but the ability to preserve the upstream links is nice.

Reply:
FWIW, course and library exports both have their context key data encoded into the to attributes of the root document (so `course.xml` or `library.xml`).

So for instance, the root `course.xml` for `course-v1:edX+DemoX+Demo_Course` would look like:

```xml
<course url_name="Demo_Course" org="edX" course="DemoX"/>
```

A long time ago (back in the XML-only days), this mattered more because we'd use it to encode multiple course runs into the same course git repo.

So we have it, but we just ignore it on import because the import mechanism is only available within the context of a specific library or course. I agree that a workflow for copying things directly without first creating a shell library could be useful. I still think that having cross-instance library references is worth looking into though. It's harder to do well, but I think it follows more closely to the workflow that people actually want to utilize.

All of which can wait until after Redwood. But yeah, let's definitely encode the full context key info into the export.

Reply:
FYI the management commands to import courses and libraries still look at [course.xml](https://github.com/openedx/openedx-demo-course/blob/master/demo-course/course/course.xml) and [library.xml](https://github.com/openedx/openedx-demo-course/blob/master/demo-content-library/library/library.xml) to determine the target key ([example invocation in Tutor](https://github.com/overhangio/tutor/blob/0e37449f40660b45aaa7539c97de1f74b40c1abb/tutor/commands/jobs.py#L173)--notice how no target key is provided).

Reply:
TIL :)

Reply:
@ormsbee @bradenmacdonald 

Coming back to this consideration about upstreams and inheritance interacting... oh, this is interesting.

Even ignoring the legacy LibraryContentBlock weirdness that we want to get rid of... if we wanted fields to be inheritable *and* upstream-defaultable, then I think we'd need to handle it something like this (psuedo-python):
```python
def get_value(block, field) -> Value|UNSET:
  block_value = block[field]                           # Priority 1: value on the block
  if block_value is not UNSET:
    return block_value
  if field.is_inheritable and block.has_parent:        # Priority 2: inherited value
     parent_value = get_value(block.parent, field)
     if parent_value is not UNSET:
        return parent_value
  return get_default(block, field)                     # Priority 3: computed default value (below)

def get_default(block, field) -> Value|UNSET:
  if field.is_inheritable:
    parent_default = get_default(block.parent, field)  # Priority 3a: inherited default
    if parent_default is not UNSET:
        return parent_default 
  if block.has_upstream:                               # Priority 3b: default from upstream
    upstream_value = block.upstream_values[field]
    if upstream_value is not UNSET:
      return upstream_value
  return field.default                                 # Priority 3c: platform-defined field default
```

> I'm also going to go out on a limb and say that upstream.{} tags are not inheritable. 😛

I like this in theory, but what are the implications? Is it that we forbid any field on InheritanceMixin from being set in a content library, and we filter out InheritanceMixin fields when copying library-defined settings into a block's `upstream.{}` defaults? Will that work out OK when we have units and subsections defined in content libraries? And is that backwards-compatible with today's LibraryContentBlock defaults?

Don't get me wrong-- I *hate* field inheritance :) I can't wait for the day where this automagical inheritance is replaced with explicit business logic in Learning Core. I just want to make sure don't introduce new edge cases in the meantime.